### DISCUSSION TITLE
How to migrate from 6.4.x to 7.0.0?

### DISCUSSION THREAD

**Discussion by cagix** [McShelby/hugo-theme-relearn#927] [Q&A]
Thank you very much for your great work on this theme. I've been using 6.4.1 up to now and after upgrading to 7.0.0 my previous archetypes no longer work, probably some other things are not working either. I see the indication of breaking changes in the release notes, but I can't really figure out what I need to adjust. Can you please provide guidance on what needs to be changed for the new version?

**Comment by McShelby** [comment]
I must admit that there are no clear advises just the hint to https://mcshelby.github.io/hugo-theme-relearn/configuration/customization/designs/index.html in the release notes.

I've taken a peak in your repo and I think (haven't tested it) you have to change as follows:

- In the Front Matter of your pages, change the `archetype` front matter to `type` (value stays the same if you don't change your _page design_ / _archetype_ naming)
- move your files `layouts/partials/archetypes/lecture-bc/article.html` to `layouts/lecture-bc/views/article.html`

**Reply by cagix** [reply]
ah. i tested the first part (`s/archetype/type/`) ( to no avail), but missed the second part. thanks for pointing that out.

**Reply by cagix** [reply]
hmmm. there seems to be more. did also the second point, and now hugo is yelling at me. it searches just in the relearn-theme location, but not also in my "theme" where the archetypes are defined (as it did before). seems, my config needs some changes, too.

**Reply by cagix** [reply]
@McShelby so i copied all my archetypes into the main theme folder (i.e. the folder where relearn is 'installed'). now hugo does find my archetypes, and i get a lot of errors of this kind: 

```
ERROR render of "section" failed: "/usr/local/share/pandoc/hugo/hugo-theme-relearn/layouts/_default/list.html:5:5": execute of template failed: template: _default/list.html:5:5: executing "body" at <.Render>: error calling Render: failed to execute template _default/views/body.html: "/usr/local/share/pandoc/hugo/hugo-theme-relearn/layouts/_default/views/body.html:1:4": execute of template failed: template: _default/views/body.html:1:4: executing "_default/views/body.html" at <partial "bodys/single.html" .>: error calling partial: "/usr/local/share/pandoc/hugo/hugo-theme-relearn/layouts/partials/bodys/single.html:2:4": execute of template failed: template: partials/bodys/single.html:2:4: executing "partials/bodys/single.html" at <.Render>: error calling Render: "/pandoc/temp/01-lexing/antlr-lexing/_index.md:1:1": failed to execute template lecture-cg/views/article.html: "/usr/local/share/pandoc/hugo/hugo-theme-relearn/layouts/lecture-cg/views/article.html:1:13": execute of template failed: template: lecture-cg/views/article.html:1:13: executing "lecture-cg/views/article.html" at <.page>: can't evaluate field page in type *hugolib.pageState
```

something in the format of the archetypes seems to have changed, too?

**Reply by McShelby** [reply]
I'll take a look into it, this evening

Also, there may be some outcome for the docs.

**Reply by cagix** [reply]
No rush! I'm in the middle of term, and this might need some major adjustments on my side - I won't be able to address that for a few months.

**Comment by zeitgen** [comment]
I just gave up upgrading at all at some point, i guess it was 5.23.2.
Maybe it's just how Hugo works in general, but even following the guidelines and not modifying bundled templates doesn't help. Custom templates still rely on bundled, which get rewritten entirely.
For a non-developer, it's just too expensive to diff each theme file figuring out what's changed while your site is down.

**Reply by McShelby** [reply]
It's perfectly fine to stay with an old version of the theme.

If you customize your templates, you've always have to expect work on upgrade.

**Reply by cagix** [reply]
> It's perfectly fine to stay with an old version of the theme.

Yep, that's true. 

However, in my experience it lasts only for a while. Eventually there will come a time when you _have_ to upgrade for some reason or other. Often it is easier to immediately follow the smaller version jumps.

**Reply by McShelby** [reply]
Personally I also prefer incremental upgrades.

In case you are far behind, it can be helpful to _not_ upgrade to the latest version in one swoop but step by step to versions in between. This may end in a bit more work but can be easily be partioned in more digestable chunks.

The overall idea for many theme changes is to make _everything_ configurable. So ideally you as a user would be in no need to touch any partials but solve any customization by configuration.

If you can not achieve your goals by configuration, feel free to post an issue.

**Reply by cagix** [reply]
But still, I tend to agree with @zeitgen. Particularly in the Hugo ecosystem, I've spent a frustrating amount of time over the years just to make the jump from one version to the next. There are simply too many variables in the game and both Hugo and the themes of course develop according to the ideas of the respective maintainers, and then my own customisations also enter the picture - it quickly takes a lot of time for me as a user to get my own project up and running again.

To be honest, I'm pondering whether and how I can remove some of the variables from the equation. At the moment, I need to keep four juggling balls in the air at the same time: Hugo, the theme, my customisations and, of course, my content together with the configuration. I am looking for a tool where I only have to maintain my content and where I don't have to worry about the rest. [Docsify](https://github.com/docsifyjs/docsify) and [LiaScript](https://github.com/liaScript) seem to be strong contenders, but they have serious shortcomings compared to the Relearn theme (given my use case). So, for now I'm stuck with Hugo+Relearn+customisations and need to check from version to version to discover which things stopped working this time.

**Reply by McShelby** [reply]
On the upside: I don't expect the themes structure to change that drastically in the future as with the past releases. It is now in a pretty good shape and complies to Hugo's best practices.

I agree, I have had my fair share adding to the upgrade pain with some clumsy implementation descisions in the first year of the fork. Those were finally removed with version 7.x. We all learn as we move along.

And for Docsify, LiaScript or any other solution: The gras is always greener on the other side. :smile:

From 25+ years experience in software development I doubt that there are fewer hassle when upgrading those projects. They may only occur fewer times due to slower release cycles (which can be pretty fast paced with Hugo and the theme)

But don't get me wrong, I don't want to sound self-convinced. I hear you!

**Reply by cagix** [reply]
Don't get me wrong - your theme is awesome and you're doing a great job! For me it's mainly Hugo with a combination of  being somewhat poorly documented and moving at great pace and me having lots of documents to maintain. Things tend to break too often in the release cycle.

**Reply by zeitgen** [reply]
I can only second @cagix: the theme is a lifesaver and literally gives me a job now =) and i'm happy we share some common ground here. 
Given the theme gets more an more stable, i might try an opposite approach to upgrading: just throw in all the latest and greatest, and then rebuild my customizations on top of that.

**Comment by McShelby** [comment]
In #980 I have greatly expanded the documentation for [output formats](https://mcshelby.github.io/hugo-theme-relearn/configuration/customization/outputformats/index.html#migration-to-relearn-7-or-higher) and [page designs](https://mcshelby.github.io/hugo-theme-relearn/configuration/customization/designs/index.html#migration-to-relearn-7-or-higher) (formerly known as archetypes) and included migration notes.

The migration instructions are linked from the [releasenotes of 7.0.0](https://mcshelby.github.io/hugo-theme-relearn/introduction/releasenotes/7/0/index.html#breaking). Give it a try! Feedback welcome!

@cagix I still haven't tested those instructions with your repo. So there might be coming some additions after I gained insights about your installation.

**Reply by cagix** [reply]
@McShelby thx a lot! after the exams end of january i'll have a look and give it a try. 

happy new year 🥳

**Reply by McShelby** [reply]
Once you are working on this again, could you create a reproducible example? I am struggling with the (at least for me) rather complex setup of your installation.

**Reply by cagix** [reply]
Unfortunately, a few projects got in the way and I didn't manage to take care of this in February/March. I think you can close the ticket. I'll just keep using the old version, and as soon as I find time, I'll change my tooling and remove my own customisations for Hugo/the theme.

**Reply by McShelby** [reply]
Understandable. No need to hurry on your side. You know where you receive help 😄 Just come back, once you're working on this again.
